### DISCUSSION TITLE
PIP-10: Fast agreement consensus protocol

### DISCUSSION THREAD

**Discussion by ja7ad** [pactus-project/PIPs#36] [PIPs]
PIP-10: [Fast agreement consensus protocol](https://pips.pactus.org/PIPs/pip-10)

**Comment by b00f** [comment]
The draft implementation is available [here](https://github.com/pactus-project/pactus/pull/837).

**Comment by b00f** [comment]
Based on the evaluation of the current Testnet, less than 30% of blocks have `3f+1` signing power from the committee members. We have [improved](https://github.com/pactus-project/pactus/pull/746/files) the consensus implementation to collect votes both before and after finalizing the agreement, ensuring fewer absentees in the vote set. There are several explanations for this:
1. Many users deploy Pactus on very basic cloud platforms or even deploy several nodes on one machine, which causes the nodes to become slow.
2. Many users are running Pactus behind a NAT; therefore, their internet connection is not stable enough. They may receive proposals too late.
3. Some nodes encounter faults during the syncing process (It is rare, but still can happens).

We can improve the fast agreement protocol by querying the votes once we have `2f+1` valid votes in the prepare phase, but this increases network bandwidth. We might even consider creating a second timer that, if expired, would return us to the pre-commit phase. However, this may not be very helpful.

The only feasible way seems to be increasing the number of fault thresholds to `5f+1`. This is considerable, but we need to think more about it and also check the security of the committee. Interestingly, the number of blocks that have `4f+1` power is above 95%. This sounds promising for the new protocol. Hopefully I didn't make any mistake on my calculations.

**Comment by b00f** [comment]
There is a situation that might break the safety of the protocol. Imagine validators $v_1$, $v_2$, $v_3$, and $v_4$. $v_1$ receives all votes from the others, but $v_2$, $v_3$, and $v_4$ only have their own votes. $v_1$ commits to block $b_1$, while the other validators move to the change proposer phase and may commit to block $b_2$.

This issue can be addressed by pre-processing messages when a validator is prepared but lacks the prepare certificates. In this case, they should share their initial value $1$ along with their prepare vote (if any).
Once a validator collects enough prepares, it should send a pre-vote with a value of 0 and include a prepare certificate.
